File: atoci/Arrays & Hashing/cpp/isAnagram.cpp
#include<iostream>
#include<unordered_map>
#include<vector>
#include<string> 

using namespace std; 

bool isAnagram(const string& s, const string& t) {
    unordered_map<char, int> mymap;

    if (s.length() != t.length()) {
        return false; 
    }

    for (char c : s) {
        mymap[c] = 0;
    }

    for (char c : s) {
        mymap[c]++;
    }

    for (char c : t) {
        if (mymap.find(c) != mymap.end()) {
            mymap[c]--;
        } else {
            return false;
        }
    } 

    for (auto& pair : mymap) {
        if (pair.second != 0) {
            return false;
        }
    }

    return true; 
    
}


int main() {

    cout << isAnagram("fag","gaf") << endl;
    cout << isAnagram("fago", "gafa") << endl; 
    return 0;
}

File: atoci/Arrays & Hashing/typescript/isAnagram.ts
export function isAnagram(s: string, t: string): boolean {

    if (s.length != t.length) {
        return false
    }

    const mymap: {[key: string]: number} = {};

    for (let index = 0; index < s.length; index++) {
        mymap[s[index]] = 0;
    }

    for (let index = 0; index < s.length; index++) {
        mymap[s[index]]++;
    }
    
    for (let index = 0; index < t.length; index++) {
        if (t[index] in mymap) {
            mymap[t[index]]--;
        } else {
            return false;
        }
    }

    for (const key in mymap) {
        if (mymap[key] != 0) {
            return false;
        }
    }

    return true; 
};

console.log(isAnagram('fag','gaf'))
console.log(isAnagram('faga','gafa'))
console.log(isAnagram('fago','gafa'))